src/core/Akka.Persistence/Journal/MemoryJournal.cs

```diff
@@ -18,73 +17,259 @@
+                {
+                    pidEvents = new List<IPersistentRepresentation>();
+                    EventsByPersistenceId[timestamped.PersistenceId] = pidEvents;
+                }
+                pidEvents.Add(timestamped);
+            }
+            finally
+            {
+                Lock.ExitWriteLock();
+            }
+
+            // Return view of all messages as LinkedList per persistence ID for API compatibility
+            Lock.EnterReadLock();
+            try
+            {
+                return EventsByPersistenceId.ToDictionary(
+                    kvp => kvp.Key,
+                    kvp => new LinkedList<IPersistentRepresentation>(kvp.Value));
+            }
+            finally
+            {
+                Lock.ExitReadLock();
+            }
+        }
+
+        /// <summary>
+        /// Delete a message and return all remaining messages.
+        /// Public API for compatibility with existing code.
+        /// </summary>
+        public IDictionary<string, LinkedList<IPersistentRepresentation>> Delete(string pid, long seqNr)
+        {
+            Lock.EnterWriteLock();
+            try
+            {
+                var currentDeleted = DeletedTo.GetValueOrDefault(pid, 0L);
+                DeletedTo[pid] = Math.Max(currentDeleted, seqNr);
+            }
+            finally
+            {
+                Lock.ExitWriteLock();
+            }
+
+            // Return view of non-deleted messages as LinkedList per persistence ID for API compatibility
+            // Use index instead of scanning entire event log
+            Lock.EnterReadLock();
+            try
+            {
+                return EventsByPersistenceId.ToDictionary(
+                    kvp => kvp.Key,
+                    kvp => new LinkedList<IPersistentRepresentation>(
+                        kvp.Value.Where(e => e.SequenceNr > DeletedTo.GetValueOrDefault(kvp.Key, 0L))));
+            }
+            finally
+            {
+                Lock.ExitReadLock();
+            }
+        }
+
+        /// <summary>
+        /// Read messages for a persistence ID within sequence range.
+        /// </summary>
+        public IEnumerable<IPersistentRepresentation> Read(string pid, long from, long to, long max)
+        {
+            Lock.EnterReadLock();
+            try
+            {
+                // Use index for O(events_for_entity) instead of O(total_events)
+                if (!EventsByPersistenceId.TryGetValue(pid, out var pidEvents))
+                    return Array.Empty<IPersistentRepresentation>();
+
+                var deletedToSeq = DeletedTo.GetValueOrDefault(pid, 0L);
+
+                return pidEvents
+                    .Where(e => e.SequenceNr > deletedToSeq
+                             && e.SequenceNr >= from
+                             && e.SequenceNr <= to)
+                    .Take(max > int.MaxValue ? int.MaxValue : (int)max)
+                    .ToArray(); // Materialize under lock
```

Comment:
For all read operations, we snapshot a collection in order to avoid funniness during after-the-fact writes